### PR TITLE
#2459 store and restore IDE state: opened files, opened part and their size

### PR DESCRIPTION
**Pull request by evidolob**
### What does this PR do?

Provide new API for components which can store their state.
Implementation for EditorAgent which now can restore openned files, scroll and cursor position,  including split editor and their sizes and position.
Implementation for Workspace(IDE UI) which can restore opened parts and thei size. 

ezgif com-optimize 9
### What issues does this PR fix or reference?
#2459

### CODE REVIEW

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/editor/texteditor/TextEditor.java [hunk 1]
```diff
@@ -132,6 +132,18 @@
...
 
     /**
+     * Returns the top visible line. Used to determine editor vertical scroll position
+     * @return the top visible line
+     */
+    int getTopVisibleLine();
+
+    /**
+     * Set (scroll) top visible line
+     * @param line the top line
```
Comment (azatsarynnyy) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/editor/texteditor/TextEditor.java]: javadoc isn't formatted

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/editor/EditorRegistry.java [hunk 2]
```diff
@@ -61,4 +61,14 @@
      */
     @NotNull
     List<EditorProvider> getAllEditorsForFileType(@NotNull FileType fileType);
+
+    /**
+     * Search editor provider by his id.
+     * Should return default editor provider if provider with that id cannot be found.
+     *
+     * @param id the provider id
```
Comment (azatsarynnyy) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/editor/EditorRegistry.java]: javadoc isn't formatted

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/component/StateComponent.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+     *
+     * @return the JSON object that represent state of the component.
+     */
+    @NotNull
+    JsonObject getState();
+
+    /**
+     * Calls when component should restore his state.
+     *
+     * @param state the component state object
```
Comment (azatsarynnyy) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/component/StateComponent.java:42]: javadoc isn't formatted

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/parts/EditorPartStack.java [hunk 4]
```diff
@@ -92,4 +93,11 @@
      */
     @Nullable
     EditorPartPresenter getLastClosed();
+
+    /**
+     * Get all parts, opened in this stack.
+     * If part doesn't have constrain for part, map should contain null value.
```
Comment (azatsarynnyy) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/parts/EditorPartStack.java]: `map should contain null value`
What map? There is no map )

File: ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/parts/PartStack.java [hunk 5]
```diff
@@ -87,4 +88,11 @@
      * Update part stack reference
      */
     void updateStack();
+
+    /**
+     * Get all parts, opened in this stack.
+     * If part doesn't have constrain for part, map should contain null value.
```
Comment (azatsarynnyy) [inline ide/che-core-ide-api/src/main/java/org/eclipse/che/ide/api/parts/PartStack.java]: `map should contain null value`
What map? There is no map )

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/editor/multipart/SplitEditorPartView.java [hunk 6]
```diff
@@ -58,4 +59,11 @@
 
     /** Removes this view from its parent widget */
     void removeFromParent();
+
+    /**
+     * Get editor multi part stack state
+     * @param splitEditorParts split editor part view mapped to their part stack
```
Comment (azatsarynnyy) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/part/editor/multipart/SplitEditorPartView.java]: javadoc isn't formatted

File: ide/che-core-ide-app/src/test/java/org/eclipse/che/ide/editor/EditorAgentStateTest.java [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+        verify(editorRegistry).findEditorProviderById("cheDefaultEditor");
+        verify(editorProvider).getEditor();
+        verify(editorMultiPartStack).createFirstPartStack();
+        verify(editorPartStack).addPart(editorPartPresenter);
+
+    }
+
+
+    @Test
+    @Ignore
```
Comment (azatsarynnyy) [inline ide/che-core-ide-app/src/test/java/org/eclipse/che/ide/editor/EditorAgentStateTest.java]: new test which is ignored?
strange

File: ide/che-core-ide-app/src/test/java/org/eclipse/che/ide/editor/EditorAgentStateTest.java [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+        when(fileTypeRegistry.getFileTypeByFile(Matchers.<VirtualFile>any())).thenReturn(fileType);
+        when(editorRegistry.getEditor(fileType)).thenReturn(editorProvider);
+        when(editorProvider.getEditor()).thenReturn(editorPartPresenter);
+        when(appContext.getWorkspaceRoot()).thenReturn(wsRoot);
+
+    }
+
+
+    @Test
+    @Ignore
```
Comment (azatsarynnyy) [inline ide/che-core-ide-app/src/test/java/org/eclipse/che/ide/editor/EditorAgentStateTest.java]: Why this new test is ignored?

File: ide/che-core-ide-app/src/test/java/org/eclipse/che/ide/workspace/perspectives/general/AbstractPerspectivePersistenceTest.java [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+
+    @Test
+    public void shouldStoreParts() throws Exception {
+        JsonObject state = perspective.getState();
+        JsonObject partStacks = state.getObject("PART_STACKS");
+        assertThat(partStacks).isNotNull();
+    }
+
+//    @Test
+//    public void shouldStoreActivePartInPartStack() throws Exception {
```
Comment (azatsarynnyy) [inline ide/che-core-ide-app/src/test/java/org/eclipse/che/ide/workspace/perspectives/general/AbstractPerspectivePersistenceTest.java]: I think this commented code is unnecessary

File: ide/che-core-ide-app/src/test/java/org/eclipse/che/ide/workspace/perspectives/general/AbstractPerspectivePersistenceTest.java [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+    @Test
+    public void shouldNotStoreEditorPartStack() throws Exception {
+        JsonObject state = perspective.getState();
+        JsonObject partStacks = state.getObject("PART_STACKS");
+        String[] keys = partStacks.keys();
+        assertThat(keys).containsOnly("INFORMATION", "NAVIGATION", "TOOLING");
+    }
+
+//    @Test
+//    public void shouldStorePartStackSize() throws Exception {
```
Comment (azatsarynnyy) [inline ide/che-core-ide-app/src/test/java/org/eclipse/che/ide/workspace/perspectives/general/AbstractPerspectivePersistenceTest.java]: I think this commented code is unnecessary

**Review by vparfonov — APPROVED** [review]

**Comment by bmicklea** [comment]
@evidolob was this merged into M6 or M7?

**Pull request merged**